developer/frameworks/articles/dependency-injection/02-cdi-dependency-injection.markdown

```diff
@@ -0,0 +1,142 @@
+    | dependencies. If you must explicitly depend on the portlet API and CDI
+    | artifacts, add them as `compileOnly` (Gradle) or `provided` (Maven)
+    | dependencies. 
+
+2.  If your portlet WAR project isn't a Bean Portlet, add this
+    `src/main/webapp/WEB-INF/beans.xml` file to it. This file informs CDI to
+    scan the project for CDI annotations. 
+
+```xml 
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://xmlns.jcp.org/xml/ns/javaee" bean-discovery-mode="all" version="1.2"
+	   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	   xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/beans_1_1.xsd">
+	<!-- This file is necessary in order to inform CDI that scanning should occur for CDI annotations. -->
+</beans>
+```
+
+| **Note:** In a PortletMVC4Spring Portlet, this `bean.xml` replaces the
+| application context file (e.g.,
+| `src/main/webapp/WEB-INF/applicationContext.xml`). 
+
+3.  Add the
+    [`@ApplicationScoped`](https://docs.oracle.com/javaee/7/api/javax/enterprise/context/ApplicationScoped.html)
+    annotation to your portlet class. 
+
+```java 
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class MyPortlet ... {
+    ...
+}
+```
+
+4.  Make sure all concrete classes you want to make injectable have the default 
+    constructor. These classes are now CDI beans. 
+
+5.  Add a scope to each CDI bean. 
+
+| Bean Scope              | Description      |
+| ----------------------- | ---------------- |
+| [`@ApplicationScoped`](https://docs.oracle.com/javaee/7/api/javax/enterprise/context/ApplicationScoped.html) | Shares the bean's state across all users' interactions with the portlet. |
+| [`@Dependent`](https://docs.oracle.com/javaee/7/api/javax/enterprise/context/Dependent.html) | (default scope) Designates the bean to be for the client bean and share the client bean's lifecycle. |
+| [`@PortletRequestScoped`](https://docs.liferay.com/portlet-api/3.0/javadocs/javax/portlet/annotations/PortletRequestScoped.html) | Associates the bean with the portlet request. |
+| [`@PortletSessionScoped`](https://docs.liferay.com/portlet-api/3.0/javadocs/javax/portlet/annotations/PortletSessionScoped.html) | Places the bean in the portlet session. |
+| [`@RenderStateScoped`](https://docs.liferay.com/portlet-api/3.0/javadocs/javax/portlet/annotations/RenderStateScoped.html) | Stores the bean as part of the portlet's render state. **Important:** The bean must implement the `PortletSerializable` interface. |
+
+6.  Use the
+    [`@Inject`](https://docs.oracle.com/javaee/7/api/javax/inject/Inject.html)
+    annotation in a CDI bean to inject another CDI bean into it. For example,
+    this code informs @product@'s CDI bean container to inject a `GuestBook` CDI
+    bean into this `guestbook` field. 
+
+```java
+@Inject 
+private GuestBook guestbook;
+```
+
+7.  Inject any [Portlet Predefined
```

Comment:
@jhinkey @sez11a I think that there is a bit of a miscommunication here. Beans that are @Dependent scoped basically get created every time they are injected with @Inject. This means that you can inject a dependent scoped predefined bean (like the portlet name bean) like this:
```
@Inject @PortletName
private String portletName;
```
However you *shouldn't* inject that into an @ApplicationScoped or @PortletSessionScoped bean because the portlet name can only be determined during @RequestScoped. You see, in Java, a String is final and so can't be a Proxy. That prevents it from being injectable into scopes that are broader than where it comes from, which is inherently the request.

So you might ask, why did the JSR 362 Expert Group define the `@PortletName` qualifier (and other dependent-scoped predefined beans) to be `@RequestScoped` then? Well, again, since the value of the portlet name is a String (which again, is final), it can't be proxied. Everything marked as @RequestScoped must be non-final and proxiable. So the only choice we had was to make these types of predefined beans @Dependent scoped. Now, after the Spec was released and I did the Spring implementation of bean portlets for Liferay, I realized that we could have required predefined portlet beans like `@PortletName` to be [CharSequence](https://docs.oracle.com/javase/7/docs/api/java/lang/CharSequence.html) instead of String, since String implements `CharSequence`, but it was too late. The train had left the station and the spec was final.